### ISSUE TITLE
502 Bad Gateway  nginx when JSP error

### ISSUE THREAD

**Issue by erugeri** [GoogleCloudPlatform/appengine-java-vm-runtime#57]
Having a jsp calling a now deleted attribute on a object I had a 502 Bad Gateway - nginx reply from the server. What's weird is that I didn't get any error in the server logs.

example:
```jsp
<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
<c:if test="${object.deletedAttribute}">test</c:if>
```

The error should be a 500 and with a server log, don't you think?
Thanks! :)

**Comment by janbartel** [comment]
Do you have a more complete jsp example? In particular, I'm looking for the exact definition of what you mean by a "deleted attribute of an object". Do you mean that the bean class has changed an no longer has a particular property, or do you mean attributes of an implicit object (if the latter, which implicit object)?

thanks
Jan

**Comment by erugeri** [comment]
Hi Jan,

I mean the bean class has changed and its property has been deleted (no attribute + no getter). Imagine the following:
```java
public class MyObject {
//   private int deletedAttribute;
//   public int getDeletedAttribute() {
//      return deletedAttribute;
//   }
}
```

Then we set the object in the model:
```java
model.put("object", new MyObject());
```

Then we forward to the previous JSP.

**Comment by janbartel** [comment]
OK, thanks for that.

I'm testing with:

public class Counter {
}

   <jsp:useBean 
           id="counter" 
           class="com.google.appengine.demos.guestbook.Counter" scope="request" /> 
    <c:if test="${requestScope.counter.counted > 0}">
      <c:out value="${requestScope.counter.count}"/>
    </c:if>


And I get an error 500 in the browser, and server log stack traces relating to  "org.apache.jasper.JasperException: javax.el.PropertyNotFoundException: Property 'counted' not found on type com.google.appengine.demos.guestbook.Counter"

I'm guessing that nginx sees the error 500 from the jetty instance and turns it into the error 502.

I have tested this outside of docker, but I'll now test inside of docker to be sure. Can you tell me anything else about your setup that might help me reproduce this?

thanks
Jan

**Comment by erugeri** [comment]
We use managed VM Java. app engine 1.9.25, gcloud maven plugin 2.0.9.72.v20150804
We use com.sun.jersey 1.7 with javax.servlet.jsp / jsp-api 2.2
What's really surprising is that jsp errors don't show in the server logs.
Thanks for your help

**Comment by janbartel** [comment]
@erugeri I haven't been able to repro this problem.

I'm running app engine 1.9.25, gcloud maven plugin 2.0.9.72.v20150804. Using the small jsp that I posted earlier. This version of appengine in docker redirects the logs to /var/log/app_engine/STDOUT.[date].log and /var/log/app_engine/STDERR.[date].log and I can see the following (elided) stacktrace in the STDOUT log:

2015-09-09 05:55:19.190:WARN:oejs.ServletHandler:qtp652761846-15: 
org.apache.jasper.JasperException: javax.el.PropertyNotFoundException: Property 'counted' not found on type com.google.appengine.demos.guestbook.Counter
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:472)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:405)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:349)
	at org.eclipse.jetty.jsp.JettyJspServlet.service(JettyJspServlet.java:107)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
        (snip)

So in short, I just can't reproduce the problem of the exception not showing up in the log. The only thing I can suggest is that perhaps jersey is swallowing the error?  My previous comment regarding the error 500 from nginx is still valid BTW.

**Comment by janbartel** [comment]
@erugeri I'm going to close this issue as I think the error swallowing is a jersey issue, and nginx is converting the error 500 into the error 502.
